### PR TITLE
Canvas and configurable jsdom windows (For things like localStorage)

### PR DESCRIPTION
**Pull request by bbarr**
Adding support for canvas and a hook into JSDom's 'created' callback for augmenting JSDom's window object on a per needs basis.

### CODE REVIEW

**Comment by bbarr** [comment]
I am not totally clear on why some of these tests are failing. Perhaps it is because of the canvas-prebuilt package not working on Node 4.x.x?

File: src/jsdom.js [hunk 1]
```diff
@@ -29,7 +29,8 @@ function fromSource(src, options) {
             FetchExternalResources: ['script'],
             ProcessExternalResources: ['script']
         },
-        virtualConsole: jsdom.createVirtualConsole().sendTo(console)
+        virtualConsole: jsdom.createVirtualConsole().sendTo(console),
+        created: options.jsdomCreated
```
Comment (mikelambert) [inline src/jsdom.js]: Sorry, couldn't find this briefly poking around the jsdom docs/code. Could you point me to the docs/code on this param and what it does?

**Comment by mikelambert** [comment]
For context...is this pull request here a fix for a regression? Did we break canvas functionality as part of switching from PhantomJS (in 0.14.x) to JSDom (0.15.0)?

Added a comment/question on the 'created' config. (Does it relate to localstorage?)

As far as tests, Travis CI tests pass on Node 4. I think Appveyor might be Windows though (Node 4 fails, Node 6/7 didn't even attempt a run...), so maybe the canvas-prebuilt doesn't work on windows?

Can you investigate the state of it there, if it's something they can/want to fix, think should be working, etc?

File: src/jsdom.js (hunk 1 above)
Comment (bbarr) [inline src/jsdom.js]: Sure, it is a callback that is triggered on creation of the jsdom. It receives (e, window). This is how one can add on localStorage shims, etc. You can see it in action in the new test.

Uncss is currently using the "old-api" in the jsdom project, so the link is: [https://github.com/tmpvar/jsdom/blob/master/lib/old-api.js#L153](https://github.com/tmpvar/jsdom/blob/master/lib/old-api.js#L153)

**Comment by bbarr** [comment]
Good question. I honestly just discovered this project, and when trying it out, discovered errors when I analyzed a page with canvas or localStorage in use. I can't say for sure if the Phantom version just worked, or if just no one had run into the problem.

This PR really conflates two changes, and if you were amenable to exposing the jsdom created callback, I suggest I break this into separate PR's, while I look a bit more into canvas-prebuilt / possible windows issues.

**Comment by XhmikosR** [comment]
By having a very quick look at canvas-prebuilt readme, it states it supports Windows x64.

We could change AppVeyor to use x64 only if there's no other way.

**Comment by XhmikosR** [comment]
On a side note, this seems like a huge dependency just for a single thing. Not sure it's worth adding it...

**Comment by XhmikosR** [comment]
I'm hitting this when I run `grunt` on grunt-uncss repo with uncss 0.15. I'm on Windows 64-bit.

When I revert back to v0.14, it works without errors. So, this is indeed needed.

**Comment by RyanZim** [comment]
Relevant jsdom issue for localstorage: https://github.com/tmpvar/jsdom/issues/1137

---

jsdom works fine without `canvas-prebuilt` installed; you only get problems when you try to use the canvas JS API. (without `canvas-prebuilt`, `<canvas>` is treated like a `<div>`)

I question the wisdom of including `canvas-prebuilt` by default. Wouldn't it be better to just make a note in the readme to install canvas-prebuilt as a peer if your JS uses the canvas API?

Another option, not sure how good; would it be possible to simply add a mock canvas api that just calls noop functions, since the actual canvas drawing doesn't affect css AFAIK?

**Comment by bbarr** [comment]
I agree with leaving canvas-prebuilt out. It is a lot of baggage, and people can just include it as needed. I have removed from this PR (which turns tests green, surprise surprise), leaving only the `jsdomCreated` hook. Thoughts?

**Comment by XhmikosR** [comment]
@bbarr: maybe we should mention instructions in README?

File: src/jsdom.js (hunk 1 above)
Comment (XhmikosR) [inline src/jsdom.js]: @bbarr: pehaps you could update jsdom to the latest version and submit a pR?

**Comment by XhmikosR** [comment]
I personally still hit the problem with canvas even if I install canvas-prebuilt with this branch https://github.com/addyosmani/grunt-uncss/tree/canvas and do `npm i && grunt`.

File: src/jsdom.js (hunk 1 above)
Comment (XhmikosR) [inline src/jsdom.js]: nvm, I went ahead and did it myself in #324. I wonder if I should cherry pick this change in that branch too.

**Comment by RyanZim** [comment]
@bbarr @XhmikosR Is this still needed?

**Comment by XhmikosR** [comment]
@RyanZim: still an issue, see the grunt-uncss repo. Without canvas it fails that's why I have it explicitly add it in devDeps there.

**Comment by giakki** [comment]
Since we transitioned to the new jsdom API, and added the "jsdom" option, you can achieve this by using `options.jsdom.beforeParse` [(docs)](https://github.com/jsdom/jsdom#intervening-before-parsing).

**Comment by rlewkowicz** [comment]
@giakki can you give any more backdrop on this. The docs are sparse and I'll be honest, I'm not particularly versed in js. Those docs are like "hey here's this method you can run" but how does that correlate to this option? I'm not supposed to construct a method here? Do I create a method and then call it using this option? I don't understand

**Pull request closed without merging**